### PR TITLE
Extend room admin api with additional attributes

### PR DESCRIPTION
**Pull request by dklimpel**
Fixes: #6427

- Add additional attributes to room admin api
`GET /_synapse/admin/v1/rooms`
- Move all room tests to test_room.py

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
* [x] Pull request includes a [sign off](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#sign-off)
* [x] Code style is correct (run the [linters](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#code-style))

### CODE REVIEW

File: docs/admin_api/rooms.md [hunk 1]
```diff
@@ -26,6 +29,16 @@ The following fields are possible in the JSON response body:
     - `name` - The name of the room.
     - `canonical_alias` - The canonical (main) alias address of the room.
     - `joined_members` - How many users are currently in the room.
+    - `joined_local_members` - How many local users are currently in the room.
+    - `version` - Version of the room.
+    - `creator` - The `user_id` of the room creator..
+    - `encryption` - Algorithm of end-to-end encryption of messages.
+    - `federatable` -Whether users on other servers can join this room.
```
Comment (clokep) [inline docs/admin_api/rooms.md:36]:
```suggestion
    - `federatable` - Whether users on other servers can join this room.
```

File: docs/admin_api/rooms.md [hunk 2]
```diff
@@ -26,6 +29,16 @@ The following fields are possible in the JSON response body:
     - `name` - The name of the room.
     - `canonical_alias` - The canonical (main) alias address of the room.
     - `joined_members` - How many users are currently in the room.
+    - `joined_local_members` - How many local users are currently in the room.
+    - `version` - Version of the room.
+    - `creator` - The `user_id` of the room creator..
+    - `encryption` - Algorithm of end-to-end encryption of messages.
```
Comment (clokep) [inline docs/admin_api/rooms.md:35]: This should mention that `null` is returned if encryption is not enabled.

File: docs/admin_api/rooms.md [hunk 3]
```diff
@@ -13,6 +13,9 @@ The following query parameters are available:
 * `order_by` - The method in which to sort the returned list of rooms. Valid values are:
   - `alphabetical` - Rooms are ordered alphabetically by room name. This is the default.
   - `size` - Rooms are ordered by the number of members. Largest to smallest.
+  - `name`, `canonical_alias`, `joined_members`, `joined_local_members`, `version`, `creator`,
+    `encryption`, `federatable`, `public`, `join_rules`, `guest_access`, `history_visibility`,
+    `state_events` - ordered by this column
```
Comment (clokep) [inline docs/admin_api/rooms.md:18]: These should be a bit more detailed, in particular it seems that some of them sort largest first, some alphabetically, etc.

File: synapse/storage/data_stores/main/room.py [hunk 4]
```diff
@@ -349,9 +401,13 @@ def is_room_blocked(self, room_id):
...
         # returned. Thus allowing us to see if there are more events to paginate through
         info_sql = """
-            SELECT state.room_id, state.name, state.canonical_alias, curr.joined_members
+            SELECT state.room_id, state.name, state.canonical_alias, state.encryption,
+              state.guest_access, state.is_federatable, state.history_visibility,
+              state.join_rules, rooms.room_version, rooms.is_public, rooms.creator,
+              curr.joined_members, curr.current_state_events, curr.local_users_in_room
             FROM room_stats_state state
             INNER JOIN room_stats_current curr USING (room_id)
+            INNER JOIN rooms USING (room_id)
```
Comment (clokep) [inline synapse/storage/data_stores/main/room.py:410]: Any idea if there are performance impacts to joining `rooms` here?

File: synapse/storage/data_stores/main/room.py [hunk 5]
```diff
@@ -388,7 +444,17 @@ def _get_rooms_paginate_txn(txn):
...
+                        "joined_local_members": room[13],
+                        "version": room[8],
+                        "creator": room[10],
+                        "encryption": room[3],
+                        "federatable": room[5],
+                        "public": room[9],
+                        "join_rules": room[7],
+                        "guest_access": room[4],
+                        "history_visibility": room[6],
+                        "state_events": room[12],
```
Comment (clokep) [inline synapse/storage/data_stores/main/room.py:457]: Any method for the ordering of these? Seems prone to error!

File: synapse/storage/data_stores/main/room.py [hunk 6]
```diff
@@ -335,6 +348,45 @@ def is_room_blocked(self, room_id):
             # Sort alphabetically
             order_by_column = "state.name"
             order_by_asc = True
+        elif RoomSortOrder(order_by) == RoomSortOrder.NAME:
+            order_by_column = "state.name"
+            order_by_asc = True
```
Comment (clokep) [inline synapse/storage/data_stores/main/room.py:353]: Is the goal here to eventually remove the `SIZE` and `ALPHABETICAL` options and just have it be via the field names? Seems odd to have multiple choices that do the same thing.

Comment (dklimpel) [reply synapse/storage/data_stores/main/room.py:353]: It is a compromise. Remove the old choices (`SIZE`, `ALPHABETICAL`) ist not good for compatibility reason. Do not add the new choices (`NAME`, `JOINED_MEMBERS`) creates a new inconsistent selection.

File: synapse/storage/data_stores/main/room.py (hunk 5 above)
Comment (dklimpel) [reply synapse/storage/data_stores/main/room.py:457]: What would you like? Alphabetically? I can change the order in sql statement, alternatively.

File: synapse/storage/data_stores/main/room.py (hunk 4 above)
Comment (dklimpel) [reply synapse/storage/data_stores/main/room.py:410]: I do not know it. I am not a SQL specialist.

File: synapse/storage/data_stores/main/room.py (hunk 6 above)
Comment (clokep) [reply synapse/storage/data_stores/main/room.py:353]: OK, that's what I thought, just wanted to make sure we were on the same page! 👍

Comment (dklimpel) [reply synapse/storage/data_stores/main/room.py:353]: Should we mark the old one as deprecated in documentation?

File: docs/admin_api/rooms.md [hunk 7]
```diff
@@ -11,8 +11,21 @@ The following query parameters are available:
 * `from` - Offset in the returned list. Defaults to `0`.
 * `limit` - Maximum amount of rooms to return. Defaults to `100`.
 * `order_by` - The method in which to sort the returned list of rooms. Valid values are:
-  - `alphabetical` - Rooms are ordered alphabetically by room name. This is the default.
-  - `size` - Rooms are ordered by the number of members. Largest to smallest.
+  - `alphabetical` - Same like `name`. This is deprecated.
+  - `size` - Same like `joined_members`. This is deprecated.
```
Comment (clokep) [inline docs/admin_api/rooms.md:15]:
```suggestion
  - `alphabetical` - Same as `name`. This is deprecated.
  - `size` - Same as `joined_members`. This is deprecated.
```

File: docs/admin_api/rooms.md [hunk 8]
```diff
@@ -11,8 +11,21 @@ The following query parameters are available:
...
+  - `name` - Rooms are ordered alphabetically by room name. This is the default.
+  - `canonical_alias` - Rooms are ordered alphabetically by main alias address of the room.
+  - `joined_members` - Rooms are ordered by the number of members. Largest to smallest.
+  - `joined_local_members` - Rooms are ordered by the number of local members. Largest to smallest.
+  - `version` - Rooms are ordered by room version. Largest to smallest.
+  - `creator` - Rooms are ordered alphabetically by creator of the room.
+  - `encryption` - Rooms are ordered by encryption of the room.
```
Comment (clokep) [inline docs/admin_api/rooms.md:22]: This is the encryption algorithm, I think? Is it done alphabetically?

File: docs/admin_api/rooms.md [hunk 9]
```diff
@@ -11,8 +11,21 @@ The following query parameters are available:
...
+  - `federatable` - Rooms are ordered by federatable option of the room.
+  - `public` - Rooms are ordered by visibility in room list.
+  - `join_rules` - Rooms are ordered alphabetically by join rules of the room.
+  - `guest_access` - Rooms are ordered by guest access option  of the room.
```
Comment (clokep) [inline docs/admin_api/rooms.md:26]:
```suggestion
  - `guest_access` - Rooms are ordered alphabetically by guest access option of the room.
```

File: docs/admin_api/rooms.md [hunk 10]
```diff
@@ -11,8 +11,21 @@ The following query parameters are available:
...
+  - `alphabetical` - Same like `name`. This is deprecated.
+  - `size` - Same like `joined_members`. This is deprecated.
+  - `name` - Rooms are ordered alphabetically by room name. This is the default.
+  - `canonical_alias` - Rooms are ordered alphabetically by main alias address of the room.
+  - `joined_members` - Rooms are ordered by the number of members. Largest to smallest.
+  - `joined_local_members` - Rooms are ordered by the number of local members. Largest to smallest.
+  - `version` - Rooms are ordered by room version. Largest to smallest.
+  - `creator` - Rooms are ordered alphabetically by creator of the room.
+  - `encryption` - Rooms are ordered by encryption of the room.
+  - `federatable` - Rooms are ordered by federatable option of the room.
```
Comment (clokep) [inline docs/admin_api/rooms.md:23]:
```suggestion
  - `federatable` - Rooms are ordered by whether the room is federatable.
```

File: docs/admin_api/rooms.md [hunk 11]
```diff
@@ -26,6 +39,16 @@ The following fields are possible in the JSON response body:
     - `name` - The name of the room.
     - `canonical_alias` - The canonical (main) alias address of the room.
     - `joined_members` - How many users are currently in the room.
+    - `joined_local_members` - How many local users are currently in the room.
+    - `version` - Version of the room.
```
Comment (clokep) [inline docs/admin_api/rooms.md:43]: I think it is worth mentioning that this is a string, maybe:

```suggestion
    - `version` - The version of the room as a string.
```

File: docs/admin_api/rooms.md [hunk 12]
```diff
@@ -26,6 +39,16 @@ The following fields are possible in the JSON response body:
...
+    - `creator` - The `user_id` of the room creator..
```
Comment (clokep) [inline docs/admin_api/rooms.md:44]:
```suggestion
    - `creator` - The `user_id` of the room creator.
```

File: docs/admin_api/rooms.md [hunk 13]
```diff
@@ -26,6 +39,16 @@ The following fields are possible in the JSON response body:
...
+    - `encryption` - Algorithm of end-to-end encryption of messages. Is `null` if encryption is not active.
+    - `federatable` - Whether users on other servers can join this room.
+    - `public` - If rooms is public or not on the server. Visibility in room directory.
```
Comment (clokep) [inline docs/admin_api/rooms.md:47]:
```suggestion
    - `public` - Whether the room is visible in room directory.
```

File: docs/admin_api/rooms.md [hunk 14]
```diff
@@ -26,6 +39,16 @@ The following fields are possible in the JSON response body:
...
+    - `join_rules` - The type of rules used for users wishing to join this room. One of: ["public", "knock", "invite", "private"]
+    - `guest_access` - Whether guests can join the room. One of: ["can_join", "forbidden"]
+    - `history_visibility` - Who can see the room history. One of: ["invited", "joined", "shared", "world_readable"]
```
Comment (clokep) [inline docs/admin_api/rooms.md:50]:
```suggestion
    - `join_rules` - The type of rules used for users wishing to join this room. One of: ["public", "knock", "invite", "private"].
    - `guest_access` - Whether guests can join the room. One of: ["can_join", "forbidden"].
    - `history_visibility` - Who can see the room history. One of: ["invited", "joined", "shared", "world_readable"].
```

File: synapse/storage/data_stores/main/room.py [hunk 15]
```diff
@@ -58,6 +58,19 @@ class RoomSortOrder(Enum):
 
     ALPHABETICAL = "alphabetical"
```
Comment (clokep) [inline synapse/storage/data_stores/main/room.py:59]: Please add a comment above `ALPHABETICAL` saying that alphabetical and size are deprecated for name and joined_members. Thanks!

File: docs/admin_api/rooms.md (hunk 8 above)
Comment (dklimpel) [reply docs/admin_api/rooms.md:22]: Values like `m.megolm.v1.aes-sha2` or `NULL`. Yes, it done alphabetically.

Comment (clokep) [reply docs/admin_api/rooms.md:22]: Please update this comment with that information, e.g.:

> Rooms are ordered alphabetically by the end-to-end encryption algorithm.

File: synapse/storage/data_stores/main/room.py (hunk 4 above)
Comment (anoadragon453) [reply synapse/storage/data_stores/main/room.py:413]: `explain` for the previous version with some example values:

```
synapse=# explain SELECT state.room_id, state.name, state.canonical_alias, curr.joined_members
synapse-# FROM room_stats_state state
synapse-# INNER JOIN room_stats_current curr USING (room_id)
synapse-# ORDER BY state.name
synapse-# LIMIT 200
synapse-# OFFSET 5;
                                             QUERY PLAN                                             
----------------------------------------------------------------------------------------------------
 Limit  (cost=255.31..255.81 rows=200 width=90)
   ->  Sort  (cost=255.30..260.44 rows=2058 width=90)
         Sort Key: state.name
         ->  Hash Join  (cost=76.99..165.99 rows=2058 width=90)
               Hash Cond: (state.room_id = curr.room_id)
               ->  Seq Scan on room_stats_state state  (cost=0.00..83.58 rows=2058 width=86)
               ->  Hash  (cost=50.33..50.33 rows=2133 width=36)
                     ->  Seq Scan on room_stats_current curr  (cost=0.00..50.33 rows=2133 width=36)
(8 rows)

synapse=# 
```

and with the new version:

```
synapse=# explain SELECT state.room_id, state.name, state.canonical_alias, curr.joined_members,
synapse-# curr.local_users_in_room, rooms.room_version, rooms.creator,
synapse-# state.encryption, state.is_federatable, rooms.is_public, state.join_rules,
synapse-# state.guest_access, state.history_visibility, curr.current_state_events
synapse-# FROM room_stats_state state
synapse-# INNER JOIN room_stats_current curr USING (room_id)
synapse-# INNER JOIN rooms USING (room_id)
synapse-# ORDER BY rooms.room_version
synapse-# LIMIT 200
synapse-# OFFSET 5;
                                             QUERY PLAN                                             
----------------------------------------------------------------------------------------------------
 Limit  (cost=331.68..332.18 rows=200 width=162)
   ->  Sort  (cost=331.67..336.44 rows=1907 width=162)
         Sort Key: rooms.room_version
         ->  Hash Join  (cost=154.90..248.91 rows=1907 width=162)
               Hash Cond: (state.room_id = curr.room_id)
               ->  Hash Join  (cost=77.91..166.90 rows=1907 width=182)
                     Hash Cond: (state.room_id = rooms.room_id)
                     ->  Seq Scan on room_stats_state state  (cost=0.00..83.58 rows=2058 width=131)
                     ->  Hash  (cost=54.07..54.07 rows=1907 width=51)
                           ->  Seq Scan on rooms  (cost=0.00..54.07 rows=1907 width=51)
               ->  Hash  (cost=50.33..50.33 rows=2133 width=44)
                     ->  Seq Scan on room_stats_current curr  (cost=0.00..50.33 rows=2133 width=44)
(12 rows)

synapse=# 
```

So a ~30% increase on the upper bound, but this query is currently only run by the admin API, and as we're getting much more information as a result of it, I don't think this is too much of an issue.

File: tests/rest/admin/test_room.py [hunk 16]
```diff
@@ -14,16 +14,694 @@
...
+
+        _order_test("encryption", [room_id_1, room_id_2, room_id_3])
+        _order_test("encryption", [room_id_1, room_id_2, room_id_3], reverse=True)
+
+        _order_test("federatable", [room_id_1, room_id_2, room_id_3])
+        _order_test("federatable", [room_id_1, room_id_2, room_id_3], reverse=True)
+
+        _order_test("public", [room_id_1, room_id_2, room_id_3])
+        # Different sort order of SQlite and PostreSQL
+        # _order_test("public", [room_id_3, room_id_2, room_id_1], reverse=True)
```
Comment (anoadragon453) [inline tests/rest/admin/test_room.py:613]: I think this is fine as the tests usually run in SQLite, even if your Synapse is configured to run with Postgres.

**Pull request merged**